src/remap/function/ip_subnet.rs

```diff
@@ -0,0 +1,254 @@
+        let mask = String::from_utf8_lossy(&bytes);
+
+        let mask = if mask.starts_with('/') {
+            // The parameter is a subnet.
+            let subnet = parse_subnet(&mask)?;
+            match value {
+                IpAddr::V4(_) => {
+                    if subnet > 32 {
+                        return Err("subnet cannot be greater than 32 for ipv4 addresses".into());
+                    }
+
+                    ipv4_addr(get_mask_bits(subnet, 4))
+                }
+                IpAddr::V6(_) => {
+                    if subnet > 128 {
+                        return Err("subnet cannot be greater than 128 for ipv6 addresses".into());
+                    }
+
+                    ipv6_addr(get_mask_bits(subnet, 16))
+                }
+            }
+        } else {
+            // The parameter is a mask.
+            mask.parse()
+                .map_err(|err| format!("unable to parse mask: {}", err))?
+        };
+
+        Ok(Value::from(mask_ips(value, mask)?.to_string()))
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .fallible_unless(value::Kind::Bytes)
+            .merge(
+                self.subnet
+                    .type_def(state)
+                    .fallible_unless(value::Kind::Bytes),
+            )
+            .with_constraint(value::Kind::Bytes)
+    }
+}
+
+/// Parses a subnet in the form "/8" returns the number.
+fn parse_subnet(subnet: &str) -> Result<u32> {
+    let re = Regex::new(r"/(?P<subnet>\d*)").unwrap();
+    let subnet = re
+        .captures(subnet)
+        .ok_or_else(|| format!("{} is not a valid subnet", subnet))?;
+
+    let subnet = subnet["subnet"].parse().unwrap(); // The regex ensures these are only digits.
+
+    Ok(subnet)
+}
+
+/// Masks the address by performing a bitwise AND between the two addresses.
+fn mask_ips(ip: IpAddr, mask: IpAddr) -> Result<IpAddr> {
+    match (ip, mask) {
+        (IpAddr::V4(addr), IpAddr::V4(mask)) => {
+            let addr: u32 = addr.into();
+            let mask: u32 = mask.into();
+            Ok(Ipv4Addr::from(addr & mask).into())
+        }
+        (IpAddr::V6(addr), IpAddr::V6(mask)) => {
+            let mut masked = [0; 8];
+            for ((masked, addr), mask) in masked
+                .iter_mut()
+                .zip(addr.segments().iter())
+                .zip(mask.segments().iter())
+            {
+                *masked = addr & mask
+            }
+
+            Ok(IpAddr::from(masked))
+        }
+        (IpAddr::V6(_), IpAddr::V4(_)) => {
+            Err("attempting to mask an ipv6 address with an ipv4 mask".into())
+        }
+        (IpAddr::V4(_), IpAddr::V6(_)) => {
+            Err("attempting to mask an ipv4 address with an ipv6 mask".into())
+        }
+    }
+}
+
+/// Returns a vector with the left `subnet_bits` set to 1,
+/// The remaining are set to 0, to make up a total length of `bytes`.
```

Comment:
I would be very interested to know if there is a better way this function could be written.

Reply:
For v4 you could simply do `~0u32 << (32 - subnet_bits)`
v6 would be similar `~0u128 << (128 - subnet_bits)`
Both `Ipv4Addr` and `Ipv6Addr` support `from` on that integer type.

Reply:
@bruceg I knew you would have a better way! Thanks.